src/storage/v2/inmemory/storage.cpp

```diff
@@ -157,6 +157,111 @@ class PeriodicSnapshotObserver : public memgraph::utils::Observer<memgraph::util
+        }(),
+        "downstream active interleaved delta found during rapid cleanup");
+    impact_tracker.update(delta.action);
+    auto prev = delta.prev.Get();
+    switch (prev.type) {
+      case PreviousPtr::Type::NULL_PTR:
+        break;
+      case PreviousPtr::Type::DELTA:
+        if (prev.delta->timestamp->load(std::memory_order_acquire) != transaction_id) {
+          prev.delta->next.store(nullptr, std::memory_order_release);
+        }
+        break;
+      case PreviousPtr::Type::VERTEX: {
+        auto &vertex = *prev.vertex;
+        vertex.delta = nullptr;
+        if (vertex.deleted) {
+          DMG_ASSERT(delta.action == Delta::Action::RECREATE_OBJECT);
+          current_deleted_vertices.push_back(vertex.gid);
+        }
+        break;
+      }
+      case PreviousPtr::Type::EDGE: {
+        auto &edge = *prev.edge;
+        edge.delta = nullptr;
+        if (edge.deleted) {
+          DMG_ASSERT(delta.action == Delta::Action::RECREATE_OBJECT);
+          current_deleted_edges.push_back(edge.gid);
+        }
+        break;
+      }
+    }
+  }
+}
+
+/** When we have interleaved deltas, we can no longer use the shortcut of
+ * only processing deltas downstream from a "head" delta, i.e., one whose `prev`
+ * is a vertex. Instead, we have to walk upstream, following `prev` pointers
+ * until we find the vertex. Obviously, this can be costly with large delta
+ * chains, so the cost is mitigated by:
+ * - only finding the vertex when we need to find the upstream vertex from an
+ *   interleaved delta
+ * - caching any intermediate subchain "heads" we find for this transaction. In
+ *   practise, this massively reduces the amount of iterating needed to be done,
+ *   as once we come across a delta we've seen before we can quickly work out to
+ *   which vertex it belongs.
+ */
+class DeltaVertexCache {
+ public:
+  explicit DeltaVertexCache(uint64_t commit_timestamp) : commit_timestamp_(commit_timestamp) {}
+
+  Vertex *GetVertexFromDelta(Delta const *delta) {
+    auto prev = delta->prev.Get();
+    if (prev.type == PreviousPtr::Type::VERTEX) return prev.vertex;
+
+    auto const it = cache_.find(delta);
+    if (it != cache_.cend()) return it->second;
+
+    std::vector<Delta const *> discovered_subchain_heads{delta};
+
+    auto const write_to_cache = [&](auto *vertex) {
+      for (auto const *uncached : discovered_subchain_heads) cache_[uncached] = vertex;
+    };
+
+    delta = prev.delta;
+    auto delta_ts = delta->timestamp->load(std::memory_order_acquire);
+    while (true) {
+      auto current_prev = delta->prev.Get();
+      if (current_prev.type == PreviousPtr::Type::VERTEX) {
+        write_to_cache(current_prev.vertex);
+        return current_prev.vertex;
+      }
+
+      MG_ASSERT(current_prev.type == PreviousPtr::Type::DELTA, "Expected DELTA in vertex delta chain");
+
+      auto const prev_ts = current_prev.delta->timestamp->load(std::memory_order_acquire);
+      if (delta_ts != prev_ts) {
```

Comment:
1. ) Why is important that delta_ts != prev_ts? I understand you want to cache just deltas from the current txn (by comparing with the commit_ts) but why do you need this condition?
2. After write_to_cache is invoked, shouldn't we clear discovered_subchain_heads, otherwise you will write multiple times to the cache the same entry. Shouldn't be too expensive but still